### ISSUE TITLE
[BUG] `mvSlice()` causes RevBayes to exit during `mcmc` run

### ISSUE THREAD

**Issue by ms609** [revbayes/revbayes#567]
**Describe the bug**
Replacing `mvScale` with `mvSlice` causes RevBayes to exit unexpectedly during an MCMC run.

**To Reproduce**
- Download [data.txt](https://github.com/user-attachments/files/17519572/data.txt)
- Run the following:
```rb
moves = VectorMoves()

chars <- readDiscreteCharacterData("data.txt")
taxa <- chars.names()
outgroup <- clade(taxa[1])
nTaxa <- chars.size()
nEdge <- 2 * nTaxa - 3

topology ~ dnUniformTopology(taxa, outgroup = outgroup)
moves.append( mvNNI(topology, weight = nEdge/2.0, tuneTarget = 0.26) )
moves.append( mvSPR(topology, weight = nEdge/10.0, tuneTarget = 0.26) )
tree_length ~ dnGamma(shape = 2, rate = 1)
moves.append( mvSlice(tree_length) )

rel_br_lengths ~ dnDirichlet( rep(1.0, nEdge) )
moves.append( mvBetaSimplex(rel_br_lengths, weight = nEdge / 5.0, tuneTarget = 0.26) )
moves.append( mvDirichletSimplex(rel_br_lengths, weight = nEdge / 10.0, tuneTarget = 0.26) )
br_lengths := rel_br_lengths * tree_length

phylogeny := treeAssembly(topology, br_lengths)

m_morph[1] ~ dnPhyloCTMC(
    tree = phylogeny,
    Q = fnJC(5),
    type = "Standard"
)
m_morph[1].clamp(chars)

monitors = [ mnScreen(printgen = 2, tree_length, likelihood = FALSE, posterior = FALSE, prior = FALSE) ]
mymcmc = mcmc(model(phylogeny), monitors, moves)
mymcmc.run(generations = 42)
```

**Expected behavior**
RevBayes exits without displaying an error message or reporting a Seg fault after a few generations.  The run completes successfully when `mvSlice` is replaced with `mvScale`.

**Computer info**
Windows 11 with a recent development build

**Comment by bredelings** [comment]
So mvSlice(x) always checks that PDFs don't change by setting x to its _current_ value.  So, this is likely to be a bug somewhere else that mvSlice complains about.  I added some code to figure out which PDF changes, and get this:
```
probability for m_morph[1] is -12202.64911 but should be -12205.97713:  delta = 3.328013235
```
I would have thought that the `DEBUG_MCMC` checks would find the same problem for mvScale( ), so I need to check why that is not the case.

I think I need to add some infrastructure for finding bugs like this that doesn't require a recompile.  The obvious thing would be to replace `-DDEBUG_MCMC` with an "option" that could be set using the `setOption(key,value)` function.  I would lean towards representing the `debugMCMC` option as an integer where 0 means no extra checks, 1 means extra checks, and 2 or more could indicate more expensive extra checks or possible extra log messages.

**Comment by ms609** [comment]
Thanks for taking a look.  An option like this would be very valuable indeed!

**Comment by bredelings** [comment]
I am going to be very busy until next Thursday with a grant-review panel.  After that I hope to get back to this.  The slice move really is pretty nice, so it would be good to fix these likelihood bugs and allow mvSlice to be used.

**Comment by bredelings** [comment]
I'm afraid that right now mvSlice basically functions as a canary for all the likelihood-not-recalculated bugs elsewhere in rb.

**Comment by ms609** [comment]
Thanks.  Yes it sounds very valuable – both as a move, and as a canary!

**Comment by bredelings** [comment]
With #575 the errors here can be found by `--setOption debugMCMC=1` even without the slice move.  Interestingly, the errors are fixed if we call `touch` and then `keep` on modified nodes after rejecting a move.  Therefore I suspect that the real problem is `treeAssembly` -- when we propose new branch lengths and reject them we somehow fail to fully restore the original branch lengths.

If so, then this would be related to #549

**Comment by ms609** [comment]
Good shout: I'm able to use `mvSlice` without error on the #549 branch.  We should verify this still holds once #549 is merged.

**Comment by davidcerny** [comment]
This can indeed no longer be reproduced on the current tip of development ([a11fcaf](https://github.com/revbayes/revbayes/commit/a11fcaf11d866b9cd5b34398eb9ea15b36b1cd80)) now that #549 has been merged into it.

Should we perhaps add this script and dataset as a `treeAssembly` integration test?

**Comment by bredelings** [comment]
Apparently we had the same idea :-)

See `tests/test_mvSlice/scripts/treeAssembly.Rev`

**Comment by davidcerny** [comment]
Ah! I somehow missed that. Closing this, then. :)
